File: src/sentinel/api/routers/plan_review.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Query, status

from sentinel.api.deps import get_settings, get_staging_service, make_page_params_dependency
from sentinel.api.schemas.common import Page, StagedRequestReceipt
from sentinel.api.schemas.plan_review import (
    PlanApprovalIn,
    PlanApprovalOut,
    PlanDecisionIn,
    PlanDecisionLogRowOut,
    PlanRowOut,
    PlanSummaryOut,
    WorkBlockOut,
)
from sentinel.api.services import plan_review_service
from sentinel.api.services.pagination import PageParams
from sentinel.api.services.staging_service import StagingService
from sentinel.config import Settings

router = APIRouter(prefix="/v1/plan-review", tags=["plan-review"])

_row_page_params = make_page_params_dependency(default_sort="suggested_order_in_block")
_decision_page_params = make_page_params_dependency(default_sort="decision_id")


@router.get("/summary", response_model=PlanSummaryOut)
def get_plan_summary(
    planning_date: str | None = Query(None),
    settings: Settings = Depends(get_settings),
) -> PlanSummaryOut:
    return plan_review_service.get_plan_summary(settings, planning_date)


@router.get("/rows", response_model=Page[PlanRowOut])
def list_plan_rows(
    planning_date: str | None = Query(None),
    page: PageParams = Depends(_row_page_params),
    settings: Settings = Depends(get_settings),
) -> Page[PlanRowOut]:
    return plan_review_service.list_plan_rows(settings, planning_date, page)


@router.get("/rows/{target_inspection_id}", response_model=PlanRowOut)
def get_plan_row(
    target_inspection_id: str,
    planning_date: str | None = Query(None),
    settings: Settings = Depends(get_settings),
) -> PlanRowOut:
    return plan_review_service.get_plan_row(settings, target_inspection_id, planning_date)


@router.get("/work-blocks", response_model=list[WorkBlockOut])
def list_work_blocks(
    planning_date: str | None = Query(None),
    settings: Settings = Depends(get_settings),
) -> list[WorkBlockOut]:
    return plan_review_service.list_work_blocks(settings, planning_date)


@router.get("/decisions", response_model=Page[PlanDecisionLogRowOut])
def list_decisions(
    planning_date: str | None = Query(None),
    page: PageParams = Depends(_decision_page_params),
    settings: Settings = Depends(get_settings),
) -> Page[PlanDecisionLogRowOut]:
    return plan_review_service.get_plan_decision_log(settings, planning_date, page)


@router.post("/decisions", response_model=StagedRequestReceipt, status_code=status.HTTP_201_CREATED)
def submit_decision(
    payload: PlanDecisionIn,
    settings: Settings = Depends(get_settings),
    staging: StagingService = Depends(get_staging_service),
) -> StagedRequestReceipt:
    return plan_review_service.stage_plan_decision(settings, payload, staging)


@router.get("/approval", response_model=PlanApprovalOut)
def get_plan_approval(
    planning_date: str | None = Query(None),
    settings: Settings = Depends(get_settings),
) -> PlanApprovalOut:
    return plan_review_service.get_plan_approval(settings, planning_date)


@router.post("/approve", response_model=StagedRequestReceipt, status_code=status.HTTP_201_CREATED)
def submit_approval(
    payload: PlanApprovalIn,
    settings: Settings = Depends(get_settings),
    staging: StagingService = Depends(get_staging_service),
) -> StagedRequestReceipt:
    """Stage a plan approval. Never applied here -- `sentinel approve-plan` commits it,
    running the full readiness checklist first (ADR 0049)."""
    return plan_review_service.stage_plan_approval(settings, payload, staging)


__all__ = ["router"]

File: frontend/src/api/planReview.ts
import { apiFetch, apiPost } from './client'
import type { PageParams } from './recommendations'
import type {
  Page,
  PlanApprovalIn,
  PlanApprovalOut,
  PlanDecisionIn,
  PlanDecisionLogRowOut,
  PlanRowOut,
  PlanSummaryOut,
  StagedRequestReceipt,
  WorkBlockOut,
} from './types'

/** Component 20/21 are scoped by `planning_date`, not the historical `DecisionScope`
 * (policy/fold/k) the rest of this app uses -- there is no backtest cell here, only a live
 * planning run. `planning_date` is optional everywhere below: omitting it asks the API for the
 * most recently built plan review. */

export function getPlanSummary(
  planningDate: string | undefined,
  signal?: AbortSignal,
): Promise<PlanSummaryOut> {
  return apiFetch('/v1/plan-review/summary', { planning_date: planningDate }, signal)
}

export function listPlanRows(
  planningDate: string | undefined,
  page: PageParams,
  signal?: AbortSignal,
): Promise<Page<PlanRowOut>> {
  return apiFetch('/v1/plan-review/rows', { planning_date: planningDate, ...page }, signal)
}

export function getPlanRow(
  targetInspectionId: string,
  planningDate: string | undefined,
  signal?: AbortSignal,
): Promise<PlanRowOut> {
  return apiFetch(
    `/v1/plan-review/rows/${encodeURIComponent(targetInspectionId)}`,
    { planning_date: planningDate },
    signal,
  )
}

export function listWorkBlocks(
  planningDate: string | undefined,
  signal?: AbortSignal,
): Promise<WorkBlockOut[]> {
  return apiFetch('/v1/plan-review/work-blocks', { planning_date: planningDate }, signal)
}

export function listPlanDecisions(
  planningDate: string | undefined,
  page: PageParams,
  signal?: AbortSignal,
): Promise<Page<PlanDecisionLogRowOut>> {
  return apiFetch('/v1/plan-review/decisions', { planning_date: planningDate, ...page }, signal)
}

/** Stages a supervisor's decision about one establishment in the proposed plan (keep_selected /
 * move_to_later_workday / do_not_proceed_as_planned). Never applied immediately -- see ADR 0049:
 * an operator later runs it through `sentinel review-plan --decisions`. Never edits Sentinel's
 * own recommendation or Component 20's geographic organization -- it only records the
 * supervisor's decision as an additional, auditable fact. */
export function submitPlanDecision(
  payload: PlanDecisionIn,
  signal?: AbortSignal,
): Promise<StagedRequestReceipt> {
  return apiPost('/v1/plan-review/decisions', payload, signal)
}

/** The latest committed approval for a planning date, if the plan has been approved. */
export function getPlanApproval(
  planningDate: string | undefined,
  signal?: AbortSignal,
): Promise<PlanApprovalOut> {
  return apiFetch('/v1/plan-review/approval', { planning_date: planningDate }, signal)
}

/** Stages a supervisor's approval of the whole plan (ADR 0049): never applied immediately --
 * an operator later runs it through `sentinel approve-plan`, which re-runs the full readiness
 * checklist (every row carries the machine recommendation, every decision has a reason, ...)
 * before writing the immutable `approved_operational_plan` artifact. */
export function submitApproval(
  payload: PlanApprovalIn,
  signal?: AbortSignal,
): Promise<StagedRequestReceipt> {
  return apiPost('/v1/plan-review/approve', payload, signal)
}
